src/algebra/direct_sum_graded.lean

```diff
@@ -0,0 +1,401 @@
+  (mul_mem : ∀ ⦃i j⦄ (gi : carriers i) (gj : carriers j), (gi * gj : R) ∈ carriers (i + j)) :
+  gcomm_monoid (λ i, carriers i) :=
+{ mul_comm := λ i j a b,
+  begin
+    have h := add_comm i j,
+    congr,
+    exact h,
+    rw subtype.heq_iff_coe_eq,
+    exact mul_comm _ _,
+    exact λ x, h.symm ▸ iff.rfl,
+  end,
+  ..gmonoid.of_submonoids carriers one_mem mul_mem}
+
+/-- Build a `gcomm_monoid` instance for a collection of `add_subgroup`s. -/
+def gcomm_monoid.of_subgroups {R : Type*} [comm_ring R] [add_comm_monoid ι]
+  (carriers : ι → add_subgroup R)
+  (one_mem : (1 : R) ∈ carriers 0)
+  (mul_mem : ∀ ⦃i j⦄ (gi : carriers i) (gj : carriers j), (gi * gj : R) ∈ carriers (i + j)) :
+  gcomm_monoid (λ i, carriers i) :=
+gcomm_monoid.of_submonoids (λ i, (carriers i).to_add_submonoid) one_mem mul_mem
+
+/-- Build a `gcomm_monoid` instance for a collection of `submodules`s. -/
+def gcomm_monoid.of_submodules {R A : Type*}
+  [comm_semiring R] [comm_semiring A] [algebra R A] [add_comm_monoid ι]
+  (carriers : ι → submodule R A)
+  (one_mem : (1 : A) ∈ carriers 0)
+  (mul_mem : ∀ ⦃i j⦄ (gi : carriers i) (gj : carriers j), (gi * gj : A) ∈ carriers (i + j)) :
+  gcomm_monoid (λ i, carriers i) :=
+gcomm_monoid.of_submonoids (λ i, (carriers i).to_add_submonoid) one_mem mul_mem
+
+end shorthands
+
+/-! ### Instances for `⨁ i, A i` -/
+
+section one
+variables [has_zero ι] [ghas_one A] [∀ i, add_comm_monoid (A i)]
+
+instance : has_one (⨁ i, A i) :=
+{ one := direct_sum.of (λ i, A i) 0 ᵍ1}
+
+end one
+
+section mul
+variables [add_monoid ι] [∀ i, add_comm_monoid (A i)] [ghas_mul A]
+
+instance : has_mul (⨁ i, A i) :=
+{ mul := λ a b, begin
+    -- the elaborator struggles here, so use tactics to assemble the expression
+    refine direct_sum.to_add_monoid (λ j,
+      direct_sum.to_add_monoid (λ i,
+        (add_monoid_hom.comp_hom (direct_sum.of A $ i + j)).comp _
+      ) a
+    ) b,
+    exact (ᵍ*),
+  end }
+
+instance : mul_zero_class (⨁ i, A i) :=
+{ mul := (*),
+  zero := 0,
+  zero_mul := by { unfold has_mul.mul, simp [direct_sum.to_add_monoid, direct_sum.of] },
+  mul_zero := by { unfold has_mul.mul, simp [direct_sum.to_add_monoid, direct_sum.of] } }
+
+instance : distrib (⨁ i, A i) :=
+{ mul := (*),
+  add := (+),
+  left_distrib := by { unfold has_mul.mul, simp, },
+  right_distrib := by { unfold has_mul.mul, simp [direct_sum.to_add_monoid, direct_sum.of], } }
+
+end mul
+
+/-! The various semiring fields are proved separately because the proofs are slow. -/
+
+section semiring
+variables [∀ i, add_comm_monoid (A i)] [add_monoid ι] [gmonoid A]
+
+private lemma one_mul (x : ⨁ i, A i) : 1 * x = x :=
+begin
+  unfold has_one.one has_mul.mul,
+  simp only [add_monoid_hom.coe_mk, to_add_monoid_of, add_monoid_hom.comp_hom_apply_apply],
+  simp only [direct_sum.to_add_monoid, dfinsupp.lift_add_hom_apply, direct_sum.of],
+  convert add_monoid_hom.congr_fun (dfinsupp.sum_add_hom_single_add_hom) x,
+  ext1 i, ext1 xi,
```

Comment:
```suggestion
  ext i xi,
```
?

Reply:
That doesn't work, but `ext i xi : 2` might work.